File: server/Models/CollaborativeFile.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel
from enum import Enum


class IWebSocketMessage(BaseModel):
    event: str
    data: dict


class CreateFileRequest(BaseModel):
    name: str


class Operation(BaseModel):
    userId: str
    type: str
    index: int
    text: str
    def apply(self, doc):
        if self.type == 'insert':
            return doc[:self.position] + self.text + doc[self.position:]
        elif self.type == 'delete':
            return doc[:self.position] + doc[self.position + len(self.text):]
        else:
            raise ValueError("Invalid operation type")

class OperationEvent(BaseModel):
    userId: str
    type: str
    index: int
    text: str
    revision: int
    def to_op(self) -> Operation:
        return Operation.model_construct(self)

class CollaborativeDocument(BaseModel):
    id: Optional[str] = None
    channelId: str
    name: str
    content: str
    operations: List[Operation]

File: client/src/@types/CollaborativeFIle.ts

export interface Content {
    [key: number]: string;
}

export interface IWebSocketMessage {
    event: string,
    data: Object,
}

export interface Operation {
    type: "insert" | "delete",
    userId: string,
    index: number
    text: string
    revision?: number
}
export interface IEditMessage extends IWebSocketMessage {
    event: 'Edit',
    data: Operation
}

export interface IChangeMessage extends IWebSocketMessage {
    event: 'change',
    data: {
        revision: number,
        operation: Operation
    }
}

export interface ISyncMessage extends IWebSocketMessage {
    event: 'sync_document',
    data: {
        revision: number,
        content: string
    }
}

export interface ICollaborativeFile {
    id: string,
    channelId: string,
    name: string,
    content: string
    //paragraphs: string[]
    //lockedParagraphs: number[]
    //edits: string[] //placeholder type
}
